source/tests/system/nidmm_driver_api_tests.cpp

```diff
@@ -0,0 +1,282 @@
+            request.mutable_vi()->set_id(GetSessionId());
+            request.set_error_code(error_status);
+            dmm::GetErrorMessageResponse response;
+
+            ::grpc::Status status = GetStub()->GetErrorMessage(&context, request, &response);
+            EXPECT_TRUE(status.ok());
+            expect_api_success(response.status());
+            return response.error_message();
+        }
+
+        ViBoolean get_bool_attribute(const char* channel_name, dmm::NiDMMAttributes attribute_id)
+        {
+            ::grpc::ClientContext context;
+            dmm::GetAttributeViBooleanRequest request;
+            request.mutable_vi()->set_id(GetSessionId());
+            request.set_channel_name(channel_name);
+            request.set_attribute_id(attribute_id);
+            dmm::GetAttributeViBooleanResponse response;
+            ::grpc::Status status = GetStub()->GetAttributeViBoolean(&context, request, &response);
+            EXPECT_TRUE(status.ok());
+            expect_api_success(response.status());
+            return response.attribute_value();
+        }
+
+        ViInt32 get_int32_attribute(const char* channel_name, dmm::NiDMMAttributes attribute_id)
+        {
+            ::grpc::ClientContext context;
+            dmm::GetAttributeViInt32Request request;
+            request.mutable_vi()->set_id(GetSessionId());
+            request.set_channel_name(channel_name);
+            request.set_attribute_id(attribute_id);
+            dmm::GetAttributeViInt32Response response;
+            ::grpc::Status status = GetStub()->GetAttributeViInt32(&context, request, &response);
+            EXPECT_TRUE(status.ok());
+            expect_api_success(response.status());
+            return response.attribute_value();
+        }
+
+        ViReal64 get_real64_attribute(const char* channel_name, dmm::NiDMMAttributes attribute_id)
+        {
+            ::grpc::ClientContext context;
+            dmm::GetAttributeViReal64Request request;
+            request.mutable_vi()->set_id(GetSessionId());
+            request.set_channel_name(channel_name);
+            request.set_attribute_id(attribute_id);
+            dmm::GetAttributeViReal64Response response;
+            ::grpc::Status status = GetStub()->GetAttributeViReal64(&context, request, &response);
+            EXPECT_TRUE(status.ok());
+            expect_api_success(response.status());
+            return response.attribute_value();
+        }
+
+    private:
+        std::shared_ptr<::grpc::Channel> channel_;
+        std::unique_ptr<::nidevice_grpc::Session> driver_session_;
+        std::unique_ptr<dmm::NiDMM::Stub> nidmm_stub_;
+        std::unique_ptr<nidevice_grpc::SessionRepository> session_repository_;
+        std::unique_ptr<dmm::NiDMMLibrary> nidmm_library_;
+        std::unique_ptr<dmm::NiDMMService> nidmm_service_;
+        std::unique_ptr<::grpc::Server> server_;
+};
+
+TEST_F(NiDMMDriverApiTest, NiDMMSelfTest_SendRequest_SelfTestCompletesSuccessfully)
```

Comment:
The code for the action step of the test implementation is `GetStub()->SelfTest`, which doesn't match the `SendRequest` "when" clause in the test name. Having just a "when" and "then" clause might be appropriate here: `SelfTest_CompletesSuccessfully`. In the future, if you end up testing `SelfTest` with an invalid session, then you can name this `ValidSession_SelfTest_CompletesSuccessfully`.

The same feedback applies to the rest of the tests below.

Reply:
changed names